sled-agent/config-reconciler/src/ledger/ledgered_sled_config_versioning.rs

```diff
@@ -0,0 +1,378 @@
+            info!(
+                log,
+                "converted config read from ledger to version {}",
+                T::DESCRIPTION
+            );
+            Some(config)
+        }
+        Err(err) => {
+            panic!(
+                "failed to convert legered config \
+                 from version {} to version {}: {}",
+                T::Previous::DESCRIPTION,
+                T::DESCRIPTION,
+                InlineErrorChain::new(&err),
+            );
+        }
+    }
+}
+
+async fn write_converted_ledger(
+    log: &Logger,
+    paths: Vec<Utf8PathBuf>,
+    sled_config: OmicronSledConfig,
+) -> OmicronSledConfig {
+    let mut config_ledger = Ledger::new_with(log, paths.clone(), sled_config);
+
+    match config_ledger.commit().await {
+        Ok(()) => (),
+        Err(err) => {
+            // We weren't able to write the new ledger, but we were still able
+            // to _read_ it (via converting an old version). Log this failure
+            // but return the config we read; we'll try converting again the
+            // next time we run.
+            warn!(
+                log,
+                "Failed to write new sled config converted from \
+                 from older version";
+                InlineErrorChain::new(&err),
+            );
+        }
+    }
+
+    config_ledger.into_inner()
+}
+
+// Terminal type for the [`VersionConversionChain`] above. This type is
+// uninhabitable (equivalent to the `Never` / `!` type), and therefore its trait
+// implementations can all safely panic (since no instance of it can exist at
+// runtime for us to have a `self`).
+#[derive(Debug, Serialize, Deserialize)]
+enum VersionConversionChainTerminal {}
+
+impl VersionConversionChain for VersionConversionChainTerminal {
+    const DESCRIPTION: &str = "NEVER_USED_TERMINAL_STATE";
+    const IS_TERMINAL: bool = true;
+
+    type Previous = Self;
+}
+
+impl Ledgerable for VersionConversionChainTerminal {
+    fn is_newer_than(&self, _: &Self) -> bool {
+        unreachable!("terminal type is uninhabitable")
+    }
+
+    fn generation_bump(&mut self) {
+        unreachable!("terminal type is uninhabitable")
+    }
+}
+
+impl TryFrom<VersionConversionChainTerminal>
+    for v4::inventory::OmicronSledConfig
+{
+    type Error = std::io::Error;
+
+    fn try_from(
+        _: VersionConversionChainTerminal,
+    ) -> Result<Self, Self::Error> {
+        unreachable!("terminal type is uninhabitable")
+    }
+}
+
+#[cfg(test)]
+pub(super) mod tests {
+    use super::*;
+    use camino_tempfile::Utf8TempDir;
+    use camino_tempfile_ext::prelude::*;
+    use omicron_test_utils::dev;
+
+    // v4 config collected from a test system.
+    const V4_CONFIG_PATH: &str = "test-data/v4-sled-config.json";
+
+    // paths for expectorate checks
+    const EXPECTORATE_V10_CONFIG_PATH: &str =
+        "expectorate/v10-sled-config.json";
+    const EXPECTORATE_V11_CONFIG_PATH: &str =
+        "expectorate/v11-sled-config.json";
```

Comment:
Do you expect these tests to get updated for v12?

Reply:
Yeah. I realize that's kind of tedious, but it seems pretty important to at least sanity check that we can convert from all prior versions up to the current. Very open to suggestions for any way to make updating these smooth. (I believe as written, the `read_config_converts_from_older_versions()` won't compile if a new version is added without updating it to account for it. But it should be a couple lines of changes to do so.)